apiserver/admin.go

```diff
@@ -187,10 +187,34 @@ func (a *admin) login(req params.LoginRequest, loginVersion int) (params.LoginRe
 			permission.IsEmptyUserAccess(everyoneGroupUser) {
 			return fail, errors.NotFoundf("model or controller access for logged in user %q", userTag.Canonical())
 		}
-		maybeUserInfo.ControllerAccess = string(controllerUser.Access)
-		maybeUserInfo.ModelAccess = string(modelUser.Access)
-		logger.Tracef("controller user %s has %v", entity.Tag(), controllerUser.Access)
-		logger.Tracef("model user %s has %s", entity.Tag(), modelUser.Access)
+		controllerAccess := controllerUser.Access
+		modelAccess := modelUser.Access
+		if !userTag.IsLocal() {
+			// There are special access rules around remote users when there are no
+			// explicit users defined.
+
+			// If there is no explicit model user for a remote user for whom
+			// the macaroon was successfully discharged, this means that the user
+			// should not have access to the model, so is permission denied.
+			if !controllerOnlyLogin && permission.ReadAccess.GreaterModelAccessThan(modelAccess) {
+				return fail, errors.Trace(common.ErrPerm)
+			}
+
+			// It is possible that there isn't a controllerUser for the remote user,
+			// so in that situation, we allows the remote user login access.
+			if permission.LoginAccess.GreaterControllerAccessThan(controllerAccess) {
+				controllerAccess = permission.LoginAccess
```

Comment:
Discussed live: as of now, permissions are expected be defined explicitly, either for the specific user, or via the special "everyone@external" group. So if there is no controller access record for the external user, and there is no "everyone@external" permission to fall back on, then we should return a permission denied. We can loosen that strict approach if needed, but right now it's best to start as restrictive as possible. All the current workflows that I am aware of will work with this strict approach.